File: examples/sample.py
"""Kitchen sink Python sample.

Purpose: compile supported constructs and expose unsupported constructs as warnings.
"""
from __future__ import annotations

import asyncio
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any


MAX_POINTS = 1000
PREMIUM_THRESHOLD = 100000
audit_counter = 0


@dataclass(frozen=True)
class Purchase:
    price: int
    category: str = "general"


@dataclass(frozen=True)
class BenefitResult:
    points: int
    message: str
    level: str
    tags: tuple[str, ...] = field(default_factory=tuple)


class AuditLog:
    """Plain class sample with methods."""

    def __init__(self) -> None:
        self.entries: list[str] = []

    def add(self, message: str) -> None:
        self.entries.append(message)


def calculate_user_benefit(
    user: dict[str, Any],
    purchases: list[Purchase],
    dry_run: bool = False,
    *tags: str,
) -> dict[str, Any]:
    """Supported-heavy function: guards, loops, data transforms, try/except/finally and condition chains."""
    if user["status"] != "ACTIVE":
        raise ValueError("inactive user")

    total_amount = 0
    for purchase in purchases:
        total_amount += purchase.price

    try:
        config_text = Path("config.json").read_text(encoding="utf-8")
        print(config_text)
    except OSError as error:
        print(error)
        raise error
    finally:
        print("benefit calculation finished")

    if dry_run:
        return {"points": 0, "message": "dry run", "level": "basic", "tags": tags}
    elif total_amount >= PREMIUM_THRESHOLD and user["rank"] == "Gold":
        return {"points": MAX_POINTS, "message": "premium benefit", "level": "premium", "tags": tags}
    elif total_amount >= 50000:
        return {"points": 500, "message": "standard benefit", "level": "standard", "tags": tags}
    else:
        return {"points": 100, "message": "basic benefit", "level": "basic", "tags": tags}


def inspect_unsupported_flow(user: dict[str, Any], values: list[int]) -> int:
    """Unsupported-heavy function: these constructs should be visible as extraction warnings today."""
    index = 0
    total = 0

    while index < len(values):
        total += values[index]
        index += 1

    match user.get("status"):
        case "ACTIVE":
            total += 10
        case "BANNED":
            total -= 100
        case _:
            total += 0

    with open("audit.log", "a", encoding="utf-8") as file:
        file.write("checked\n")

    first, *rest = values
    normalized = [value * 2 for value in rest if value > first]
    label = f"rank:{user.get('rank', 'None')}" if total > 0 else "none"
    callback = lambda value: value + 1
    print(label)

    return sum(callback(value) for value in normalized) + total


async def fetch_remote_points(user_id: str) -> int:
    await asyncio.sleep(0)
    return len(user_id)


def stream_points(values: list[int]):
    for value in values:
        yield value

File: examples/sample.ts
// Kitchen sink TypeScript sample.
// Purpose: compile supported constructs and expose unsupported constructs as warnings.

import { readFile } from "fs/promises";

const MAX_POINTS = 1000;
const PREMIUM_THRESHOLD = 100000;
let auditCounter = 0;

type UserStatus = "ACTIVE" | "INACTIVE" | "BANNED";
type BenefitLevel = "premium" | "standard" | "basic";

interface Purchase {
  price: number;
  category?: string;
}

interface User {
  id: string;
  status: UserStatus;
  rank: string;
  purchaseHistory: Purchase[];
  metadata?: Record<string, unknown>;
}

interface BenefitResult {
  points: number;
  message: string;
  level: BenefitLevel;
}

class AuditLog {
  private entries: string[] = [];

  add(message: string): void {
    this.entries.push(message);
  }
}

/**
 * Supported-heavy function: guards, loops, data transforms, try/catch/finally and condition chains.
 */
async function calculateUserBenefit(
  user: User,
  options: { configPath?: string; dryRun?: boolean } = {},
  ...tags: string[]
): Promise<BenefitResult> {
  if (user.status !== "ACTIVE") {
    throw new Error("inactive user");
  }

  let totalAmount = 0;
  for (const history of user.purchaseHistory) {
    totalAmount += history.price;
  }

  for (let i = 0; i < tags.length; i++) {
    console.log(tags[i]);
  }

  try {
    const configText = await readFile(options.configPath ?? "config.json", "utf8");
    console.log(configText);
  } catch (error) {
    console.error(error);
    throw error;
  } finally {
    auditCounter += 1;
    console.log("benefit calculation finished");
  }

  if (options.dryRun) {
    return { points: 0, message: "dry run", level: "basic" };
  } else if (totalAmount >= PREMIUM_THRESHOLD && user.rank === "Gold") {
    return { points: MAX_POINTS, message: "premium benefit", level: "premium" };
  } else if (totalAmount >= 50000) {
    return { points: 500, message: "standard benefit", level: "standard" };
  } else {
    return { points: 100, message: "basic benefit", level: "basic" };
  }
}

/**
 * Unsupported-heavy function: these constructs should be visible as extraction warnings today.
 */
function inspectUnsupportedFlow(user: User, values: number[]): number {
  let index = 0;
  let total = 0;

  while (index < values.length) {
    total += values[index];
    index += 1;
  }

  do {
    total -= 1;
  } while (total > 1000);

  switch (user.status) {
    case "ACTIVE":
      total += 10;
      break;
    case "BANNED":
      total -= 100;
      break;
    default:
      total += 0;
  }

  for (const key in user.metadata ?? {}) {
    console.log(key);
  }

  const [first = 0, ...rest] = values;
  const { rank = "None" } = user;
  const normalized = rest.map((value) => value * 2).filter((value) => value > first);
  const label = total > 0 ? `rank:${rank}` : "none";
  console.log(label);

  return normalized.reduce((sum, value) => sum + value, total);
}

const arrowBenefit = (value: number): number => (value > 0 ? value : 0);

